### ISSUE TITLE
Disable logical operators (or fix)

### ISSUE THREAD

**Issue by jsmentch** [neuroscout/neuroscout#802]
What is the expected behavior of the And, Or, Not transformations? If I have 2 predictors and I apply And to both of them like so:

<img width="299" alt="Screen Shot 2020-07-20 at 4 10 03 PM" src="https://user-images.githubusercontent.com/20958589/87982064-345a4580-caa4-11ea-9c00-d46cc101fe52.png">

I'd expect to get back one or both with this applied, but in the design matrix in the web app it seems like no transformation was applied:
<img width="722" alt="Screen Shot 2020-07-20 at 4 09 52 PM" src="https://user-images.githubusercontent.com/20958589/87982150-5358d780-caa4-11ea-9560-aa2bc9b3d674.png">

These 2 features are almost always not on at the same time, so an And operations should leave not much signal. I may be doing something wrong but from the instructions it's not too clear and they don't seem to be mentioned specifically in the BIDS StatsModel specification referenced in the Neuroscout Docs in that section

**Comment by adelavega** [comment]
Yes the idea is for it to be a logical and operation. Looks like there's something funny going on here.

Looking at the BIDS Stats Model I see this:

```
        {
          "Name": "And",
          "Input": [
            "as-Animal",
            "as-Alarm"
          ],
          "Output": [
            "as-Animal",
            "as-Alarm"
          ]
        }
```

which indicates that both are being taken as inputs, but the output is set to both as well, which is incorrect.
It should indicate only a sign output field.

Ideally, the name of the new field would be a new one, rather than doing an inplace operation.

However, Neuroscout's frontend can't (currently) handle transformations that produce new names (only can handle in place operations). Until we add that feature, these logical transofmrations should probably be disabled. 

FWIW, in pyns you can try any of the operations supports by BIDS Stats Model. They are not wel documented as this is a DRAFT spec, but you can see the functions in the pybids code here: https://github.com/bids-standard/pybids/tree/master/bids/analysis/transformations

If you want to use the `And` operation for now I suggest using `pyns` and setting the field of `Output` to a new column name. You can then use the `Select` transformation to only keep the variable if you want to use in the final model (e.g. drop the columns that went into the `And` operation, and only keep the new variable).

**Comment by jsmentch** [comment]
Ok, got it! I haven't used pyNS to actually set up an analysis yet, just for pulling features, and I'm running into a bit of trouble figuring this out. Here is what I am attempting at the moment:

analysis = neuroscout.analyses.create_analysis(
    dataset_name='SherlockMerlin', name='merlin speaker face',
    predictor_names=['speaker_id_merlin', 'face_id_1'],
    transformations=[{'Input': ['speaker_id_merlin', 'face_id_1'], 'Name': 'And', 'Output':'speaker_face'}],
    hrf_variables=['speaker_face'], 
    subject=['19'],
    task =['4'],
)

ValueError: Task name does not match any tasks in the dataset


From what I can tell the task is '4' for Merlin and '45' for sherlock but getting this error. And do you know how I might format the "transformations" field to also including the 'select' transformation? Can't find an example using transformations.

Maybe unrelated, but I'm also unable to view my analyses in the web app at the moment, it hangs with the blue spinning circles. I can start a new analysis though through the web

**Comment by adelavega** [comment]
The function is actually trying to make this easier for you by letting pass in the name of the task, rather than the ID.

so you should be able to simply say "MerlinMovie" or whatever.

**Comment by adelavega** [comment]
Thanks, I just loaded up a dev version temporarily. We really need to set up a staging server now that we have users sorry! Should be working again

**Comment by jsmentch** [comment]
Ok, both are working, thanks. Haha, all good!

**Comment by jsmentch** [comment]
I am still having trouble using And via pyNS. If you have a second do you know what I'm doing wrong? The model does not seem to be built correctly with the new variable that is the Output of And. I just want a variable that is the And of speech and face_1 and I want to convolve that w/HRF and then have a contrast of that vs everything else.

I create an analysis like this:
```
#create an analysis
analysis = neuroscout.analyses.create_analysis(
    dataset_name='SherlockMerlin', name='logical and face speech',
    description='try to use logical and',
    predictor_names=['rot_x', 'rot_y', 'rot_z', 'trans_x', 'trans_y', 'trans_z', 'global_signal', 'csf', 'white_matter','framewise_displacement', 'face_id_1', 'speech'],
    transformations=[{'Input': ['face_id_1', 'speech'], 'Name': 'And', 'Output': 'face_speech'},
                     {'Input': 'face_speech', 'Name': 'Select'},
                     {'Input': 'face_speech', 'Name': 'Convolve'}],
    contrasts=[{"ConditionList": [
            "face_speech"
          ],
          "ContrastType": "t",
          "Name": "face_speech",
          "Weights": [
            1
          ]}],
    #hrf_variables=['face_speech'],   #ValueError: HRF Variables must be a subset of all variables
    #subject=['19'],
    task ='MerlinMovie'
)
```

analysis.model seems to be missing some things related to the new face_speech variable:
```
{'Input': {'Subject': ['20',
   '36',
   '25',
   '29',
   '21',
   '23',
   '22',
   '24',
   '26',
   '28',
   '30',
   '31',
   '33',
   '34',
   '19',
   '35',
   '32',
   '27'],
  'Task': 'MerlinMovie'},
 'Name': 'logical and face speech',
 'Steps': [{'Contrasts': [{'ConditionList': ['face_speech'],
     'ContrastType': 't',
     'Name': 'face_speech',
     'Weights': [1]}],
   'DummyContrasts': {'Type': 't'},
   'Level': 'Run',
   'Model': {'X': ['rot_x',
     'rot_y',
     'rot_z',
     'trans_x',
     'trans_y',
     'trans_z',
     'global_signal',
     'csf',
     'white_matter',
     'framewise_displacement',
     'face_id_1',
     'speech']},
   'Transformations': [{'Input': ['face_id_1', 'speech'],
     'Name': 'And',
     'Output': 'face_speech'},
    {'Input': 'face_speech', 'Name': 'Select'},
    {'Input': 'face_speech', 'Name': 'Convolve'},
    {'Input': [], 'Name': 'Convolve'}]},
  {'DummyContrasts': {'Type': 'FEMA'}, 'Level': 'Subject'},
  {'DummyContrasts': {'Type': 't'}, 'Level': 'Dataset'}]}
```

receive an error in the report
```
report
{'generated_at': '2020-08-19T19:4',
 'result': None,
 'sampling_rate': None,
 'scale': True,
 'status': 'FAILED',
 'traceback': 'Error validating analysis.\nReplacement data has shape (495,); must have same shape as existing data (418,).',
 'warnings': []}
```

**Comment by adelavega** [comment]
I'll take a look. Your model looks reasonable. Looks like if you omit `hrf_variables` from `create_analysis` it should omit that Transformation, but it's not. That could be where the error is coming from but it's hard to tell, let me look a bit deeper.

**Comment by adelavega** [comment]
Okay, the short story is general issue is that the two variables are not aligned in time, and this causes several issues in the `And` transformation. 

Here's the long story:


First, it's important to understand that prior to any transformation, variable are "sparse". 

That is, the variables are not necessarily uniformly sampled (e.g. have a value at every TR, or at every second). Instead, it depends on the variables. For example, in the case of `speech`, there is a value for everytime there *is* speech, but not a value for when there is not. Once we convert variables to "dense" format (e.g. uniform sampling at some pre-determiened frequency, in this case `TR`), the absence of values gets converted to an explicit 0. 

In the case of the face variables (and actually most visual predictors), although they are in "sparse" format, they are actually uniformly sampled at 1hz. That is because we had the explicitly sample the Video to extract still frames to feed into the Face Detection model. If I remember correctly, we also only have values for frames in which *any* face was detected.

Details aside, what time means is that when you perform an `And` operation, you are taking two vectors which are not aligned temporally and trying to take a logical operation them. 

Specifically, pybids will just join these two variables (one of them has 418 onsets, the other 495) into a pandas dataframe, and take an `And` operation on the columns. 

In this case, this causes an error when putting the results variable back, and we should catch that error better. However, regardless even if this didnt fail, the results wouldn't necessarily make sense.

What you need to do here to make this work is align these two variables in time, and then apply the `And` operation.
One way to do this would be to resample both variables into the same frequency, and then apply logical operations.
However, resample would probably make variables such as speech have values other than just 0 and 1 (since some frames will have speech in only some part of them), so you might have to also apply a `Threshold` transformation prior to the `And` transformation.

Sorry that's so complicate but I'm happy to help you work through it, this is a useful example of where we could try to make things easier in Neuroscout.

**Comment by adelavega** [comment]
Also, I do think the specific error your getting is a pybids bug so, I'll raise that.

**Comment by jsmentch** [comment]
Oh wow!!! Alright, that makes sense. Ok so to get around this for now, sounds like I will need to download the variables I want, ensure they are densely sampled at the same rate, and then upload them as new predictors to use. I think I can manage that.

**Comment by adelavega** [comment]
Actually, there is a transformation called `ToDense` and a transformation called `Threshold`. 
You may have to play around with how finely you resample to find a good number. It's not obvious what the right number is. 

Also, this was a  general recommendation, but once the bug is fixed, you should be able to do an `And` operation on two events that are not perfectly aligned. As long as some events aligned, you would get a `1` value there and a `0` value where they don't align. I just wanted you to know about how things are sampled, because that might cause a `0` value if two events just miss (even by just a few ms), even though conceptually a `1` would make more sense.

**Comment by jsmentch** [comment]
Ah great! I will give this a try

**Comment by jsmentch** [comment]
I've tried adding the `ToDense` and `Threshold` transformations but now getting this error in the report. The lengths seem to be matched now but different shapes:

```
'traceback': 'Error validating analysis.\nReplacement data has shape (15435,); must have same shape as existing data (15435, 1).',
```

**Comment by adelavega** [comment]
Okay, that fix is incoming in a new PR to pybids. Once that is merged, I will load it into Neuroscout and it should fix this issue. 

https://github.com/bids-standard/pybids/pull/649#pullrequestreview-472071282
